MovingBlocks/Terasology
# [Breaking all] Gestalt v7 re-integration....

### Contains
 Re-integration gestaltv7

 * Changing engine module to package module. ( classpath modules dropped)
 * engine moved in self package (`org.terasology.engine` now)
 * create syntetic package module  for nui (easier then write all in whitelist)
 * Update gestalt to 7.0.5
 * Update gson to 2.8.5
 * Update TeraNui to 2.0.0
 * Update Guava to 29.0-jre
 * Update Reflections to org.terasology:reflections:0.9.12-MB (gestalt build it)
 * Reimplemented Module download (gestalt drop `BaseModule` class)
 * Restore Module loading on gestalt v7
 * Use AutoreloadingAssetManager (seems asset must reload easier now)
 * Replace condition `Module#isSourceModule` to checking Module's files source == `DirectoryFileSource` - must be equal ( affected WorldGeneratorManager)
 * add Gestalt's classes to whitelist
 * remove duplication classes from `org.terasology.input`(`org.terasology.engine.input` now). We have this classes in TeraNUI
 * move `cacheReflections` task on after `classes` task (partial resolve https://github.com/MovingBlocks/Terasology/issues/4126#issuecomment-692232675) (reflections gathering for running from sources now)
 * remove `cleanReflections` after `jar` task
 * renamed `cacheReflections`'s artifact `reflections.cache` to `manifest.json` and serializer changed to Json for using gestaltv7's default values

Fixes #2445

### Testing
1. Check annotation marked classes reloading (e.g. WorldGenerator)
   1. Run game from idea.
   2. create game, exit
   3. create dummy world generator in active module
   4. create game with advanced settings, 
   5. try found your new world generator

2. Check assets reloading
   1. Run game from idea. (in debug mode)
   2. create game
   3. change any asset of active module
   4. recompile classes 
   4. exit game and load game
   5. asset must be edited in game.

3. Check module downloading 
   1. Run game 
   2. Press New game
   3. Press Create game
   4. Press Advanced game setup
   5. try download any module.

4. Fast check for 70% sure that game works
   1. run `gradlew test`
   2. try found failed tests ;)

5. Check guava 29.0-jre :
   1. Play with MasterOfOreon ( minions uses guava's changed Futures for pathfinding)
   2. make them to use pathfinding

6.* Check what everything works fine:
   1. Play with all modules
   2. check all features
   3. if you found bug, be sure that bug from this PR :)

* - impossible

### Outstanding before merging
- [X] Engine compiles and runs. 
- [x] Reimplement in-game module downloading.
- [?] Check and fix from server module downloading.
- [x] Check and fix Assets.
- [x] Check and fix module loading.
- [x] Check and fix whitelist.
- [x] Update module space

#### Linked PRs
https://github.com/MovingBlocks/TeraNUI/pull/27
